utils/core_maths/src/lib.rs

```diff
@@ -0,0 +1,1142 @@
+    /// let g = ((e * e) + 1.0) / (2.0 * e);
+    /// let abs_difference = (f - g).abs();
+    ///
+    /// // Same result
+    /// assert!(abs_difference < 1.0e-10);
+    /// ```
+    fn cosh(self) -> Self;
+
+    /// Hyperbolic tangent function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let e = std::f64::consts::E;
+    /// let x = 1.0_f64;
+    ///
+    /// let f = x.tanh();
+    /// // Solving tanh() at 1 gives `(1 - e^(-2))/(1 + e^(-2))`
+    /// let g = (1.0 - e.powi(-2)) / (1.0 + e.powi(-2));
+    /// let abs_difference = (f - g).abs();
+    ///
+    /// assert!(abs_difference < 1.0e-10);
+    /// ```
+    fn tanh(self) -> Self;
+
+    /// Inverse hyperbolic sine function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let x = 1.0_f64;
+    /// let f = x.sinh().asinh();
+    ///
+    /// let abs_difference = (f - x).abs();
+    ///
+    /// assert!(abs_difference < 1.0e-10);
+    /// ```
+    fn asinh(self) -> Self;
+
+    /// Inverse hyperbolic cosine function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let x = 1.0_f64;
+    /// let f = x.cosh().acosh();
+    ///
+    /// let abs_difference = (f - x).abs();
+    ///
+    /// assert!(abs_difference < 1.0e-10);
+    /// ```
+    fn acosh(self) -> Self;
+
+    /// Inverse hyperbolic tangent function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let e = std::f64::consts::E;
+    /// let f = e.tanh().atanh();
+    ///
+    /// let abs_difference = (f - e).abs();
+    ///
+    /// assert!(abs_difference < 1.0e-10);
+    /// ```
+    fn atanh(self) -> Self;
+}
+
+impl CoreFloat for f32 {
+    #[inline]
+    fn floor(self) -> Self {
+        libm::floorf(self)
+    }
+
+    #[inline]
+    fn ceil(self) -> Self {
+        libm::ceilf(self)
+    }
+
+    #[inline]
+    fn round(self) -> Self {
+        libm::roundf(self)
+    }
+
+    #[inline]
+    fn round_ties_even(self) -> Self {
+        libm::rintf(self)
+    }
+
+    #[inline]
+    fn trunc(self) -> Self {
+        libm::truncf(self)
+    }
+
+    #[inline]
+    fn fract(self) -> Self {
+        self - self.trunc()
+    }
+
+    #[inline]
+    fn abs(self) -> Self {
+        libm::fabsf(self)
+    }
+
+    #[inline]
+    fn signum(self) -> Self {
+        if self.is_nan() {
+            Self::NAN
+        } else {
+            1.0_f32.copysign(self)
+        }
+    }
+
+    #[inline]
+    fn copysign(self, sign: Self) -> Self {
+        libm::copysignf(self, sign)
+    }
+
+    #[inline]
+    fn mul_add(self, a: Self, b: Self) -> Self {
+        libm::fmaf(self, a, b)
+    }
+
+    fn div_euclid(self, rhs: Self) -> Self {
+        let q = (self / rhs).trunc();
+        if self % rhs < 0.0 {
+            return if rhs > 0.0 { q - 1.0 } else { q + 1.0 };
+        }
+        q
+    }
+
+    fn rem_euclid(self, rhs: Self) -> Self {
+        let r = self % rhs;
+        if r < 0.0 {
+            r + rhs.abs()
+        } else {
+            r
+        }
```

Comment:
General issue: We should minimize the amount of math code in this crate. If we must add any, it should be well tested. One way to test it is to run it against the standard library implementation on a suitably diverse set of inputs.

Specific issue: Pretty sure the div/rem euclid impls here aren't right. The [standard library impl Rem](https://doc.rust-lang.org/std/primitive.f64.html#impl-Rem%3Cf64%3E-for-f64) says:

> The remainder has the same sign as the dividend and is computed as: `x - (x / y).trunc() * y`

But rem_euclid is a different thing. It is the remainder relative to the floor, not the trunc. You can't derive that by taking `.abs()`.

Reply:
All code here is copied from `std`, with calls to intrinsics replaced by libm. You're linking to `Rem`, which is a different operation to [`rem_euclid`](https://doc.rust-lang.org/src/std/f64.rs.html#328).

Reply:
Ok `r + rhs.abs()` I think is correct. We care a lot about rem_euclid for calendrical calculations.

Reply:
`rhs.abs()` is the more general form, we use `rhs` in our version because we require `rhs > 0`